File: crates/agent/src/ending.rs
//! HOW A TURN ENDED, AS A FACT (R17-P0-2).
//!
//! A turn used to end by ABSENCE: `state.task = None`, and every summary on the
//! page then read the same absence as success. A run that abandoned a six-part
//! task reported `main finished "…"` with a `Read the reply` button, and the
//! reply it landed on was the model's own malformed tool call.
//!
//! The machine already knew the difference — the answered path, the round
//! ceiling and the stop are three different lines of `step` — so this makes the
//! ending a RECORD with a kind rather than a hole where the task was. One fact,
//! `core.ended`, emitted by the pure step function at every ending it owns; the
//! card, the board row and the conversation are folds of it (`core::failure::ending`),
//! which is the shape `core/src/failure/stopped_notice.rs` already proved for the stop.
//!
//! `core.stopped` keeps its own fact and its own file: it is emitted at the
//! boundary rather than by an arm of the machine, it carries a wording no other
//! ending needs, and a second record of it here would be the split round 16 was
//! spent removing.

use kernel::EventKind;

use crate::effect::Effect;
use crate::state::AgentState;

/// The one ending fact. Payload: `{"why": …, "rounds": n}`.
pub const ENDED: &str = "core.ended";

/// The model replied with prose. The turn's cheap exit, and the ONLY ending
/// after which there is a reply to read.
pub const ANSWERED: &str = "answered";

/// The model replied with machine output — a tool call this page could not read
/// (`reply::malformed_call`). Nothing ran and nothing was answered.
pub const NO_ANSWER: &str = "no answer";

/// The turn used every round of tool calls its agent file allows.
pub const ROUND_CEILING: &str = "round ceiling";

/// THE PASS BUDGET RAN OUT WHILE THE WORK WAS STILL GOING (22). Its own ending
/// beside `ROUND_CEILING`, and for R17-P0-2's reason: a six-part task was
/// abandoned and reported as `main finished`. A turn that stopped because it
/// ran out of laps — not because it finished — must never read as an answer,
/// so the surfaces get a word for it and the conversation says the work
/// stopped on the budget.
pub const PASS_CEILING: &str = "pass ceiling";

/// THE DECLARED GOAL WAS NOT MET (26). A `goal.check` was declared, the harness
/// ran it, and the command exited non-zero with no lap left to try again. Its
/// own ending beside `PASS_CEILING` because a surface can offer a different act
/// for it, which is the whole rule for naming one: the pass ceiling says the
/// budget ran out while the work was still moving and the act is to raise
/// `passes:`; this says a command the file itself nominated still fails, and
/// the act is to read what it printed. It is also the one ending on this page
/// that rests on nothing anybody said — not the model's summary, not a critic's
/// verdict, not a fold over which tools were called. An exit code.
pub const GOAL_UNMET: &str = "goal unmet";

/// THE REVIEWER SAID NO (25). A separate agent — `role: critic`, its own
/// prompt, its own Worker, no sight of this conversation — was handed the work
/// and did not clear it. The answer is real and is shown; what this says is
/// that the last thing to look at it disagreed with it. It is not the machine's
/// judgement of the work, and it is not this agent's own: `critique` is the
/// stage where a model marks its own homework, and this ending exists so that
/// homework marked by somebody else cannot be filed as a pass.
pub const CRITIC_FAULTED: &str = "critic faulted";

/// THE MODEL ANSWERED, AND THE ANSWER IS SHOWN — but the turn changed a file
/// and no command has run since, so nothing on this page knows whether it
/// worked (`crate::verify`). Not a failure and not a judgement about the
/// answer: an ending that says which of the two things a reader might assume is
/// actually known.
pub const UNCHECKED: &str = "unchecked";

/// A STAGE HAD NO BRIEF, SO THE TURN DID NOT TAKE IT (`crate::brief`). The
/// words a stage enters with are data now — `public/stages/<key>.md`, fetched
/// at boot — and the one thing that must never happen when a file is missing is
/// the stage running anyway on a compiled-in copy: a plan stage that writes no
/// plan still looks exactly like a plan stage that ran.
pub const BRIEF_MISSING: &str = "brief missing";

/// REFUSE THE STAGE, IN WORDS, AND END THE TURN. Two facts and not one: the
/// ending carries the KIND, which the board and the card fold, and the note
/// carries the sentence — which has to name the one file a person must add,
/// and a kind cannot say which of five it was.
pub(crate) fn unbriefed(state: &mut AgentState, key: &str) -> Vec<Effect> {
    let said = format!(
        "The {key} stage has no brief, so it was not entered and this turn stopped before \
         it started. Add {} and reload — its whole body is what that stage is told.",
        crate::brief::path_of(key)
    );
    let note = Effect::Emit {
        kind: EventKind::Custom {
            kind: "core.note".into(),
            payload_json: serde_json::to_string(&said).unwrap_or_default(),
        },
    };
    vec![note, end(state, BRIEF_MISSING)]
}

/// END THE TURN, AND SAY WHY. Every arm of `step` that ends one goes through
/// here, so "what a turn ending clears" is written once and the reason is never
/// optional — the compiler asks for it at the call site.
pub(crate) fn end(state: &mut AgentState, why: &str) -> Effect {
    let payload = serde_json::json!({ "why": why, "rounds": state.tool_rounds }).to_string();
    (state.task, state.retries, state.tool_rounds) = (None, 0, 0);
    crate::verify::clear(state); // evidence is about ONE turn, like the rounds
    Effect::Emit {
        kind: EventKind::Custom {
            kind: ENDED.into(),
            payload_json: payload,
        },
    }
}

/// Why the turn ended, out of the payload. An unreadable record says nothing
/// rather than guessing a reason, and the surfaces treat that as today's
/// behaviour — which is what every log written before this reads as.
pub fn ended_why(payload_json: &str) -> String {
    serde_json::from_str::<serde_json::Value>(payload_json)
        .ok()
        .and_then(|v| v.get("why")?.as_str().map(str::to_string))
        .unwrap_or_default()
}

/// How many rounds of tool calls it had completed when it ended.
pub fn ended_rounds(payload_json: &str) -> u16 {
    serde_json::from_str::<serde_json::Value>(payload_json)
        .ok()
        .and_then(|v| v.get("rounds")?.as_u64())
        .unwrap_or(0) as u16
}

/// Whether an effect is one of these records rather than work. `stop::boundary`
/// asks: a turn that ended on its own is not a turn you cut off.
pub(crate) fn is_ending(effect: &Effect) -> bool {
    matches!(
        effect,
        Effect::Emit { kind: EventKind::Custom { kind, .. } } if kind == ENDED
    )
}

#[cfg(test)]
mod tests {
    use crate::state::AgentState;

    /// The reason and the count survive the round trip, and a turn that ended
    /// holds nothing a running one holds.
    #[test]
    fn an_ending_carries_its_reason_and_the_rounds_behind_it() {
        let mut state = AgentState::new();
        (state.task, state.tool_rounds) = (Some("do it".into()), 7);
        let effect = super::end(&mut state, super::ROUND_CEILING);
        assert!(super::is_ending(&effect));
        let crate::effect::Effect::Emit {
            kind: kernel::EventKind::Custom { payload_json, .. },
        } = &effect
        else {
            panic!("an ending is one emitted Custom fact")
        };
        assert_eq!(super::ended_why(payload_json), super::ROUND_CEILING);
        assert_eq!(super::ended_rounds(payload_json), 7);
        assert!(state.task.is_none(), "an ended turn holds no task");
        // A log written before this fact existed reads as no reason at all.
        assert_eq!(super::ended_why("null"), "");
    }
}

File: packages/agent/src/ending.js
/**
 * HOW A TURN ENDED, AS A FACT (R17-P0-2) — and, since this rewrite, ON A
 * SIGNAL.
 *
 * A turn used to end by clearing `task`, and every surface then read that hole
 * as success: a run that abandoned a six-part task reported `main finished` with
 * a `Read the reply` button pointing at the model's own malformed tool call. So
 * an ending is a RECORD with a reason, and the card, the row and the
 * conversation are folds of it.
 *
 * WHAT CHANGED IN THE PORT. The Rust decided the reason by ABSENCE — no call
 * could be read out of the text, therefore the model answered — which made a
 * truncated reply, a refusal and a real answer one outcome, and left
 * `reply::malformed_call` guessing at the difference from the shape of the
 * prose. The provider says which of them it was, in one field, and
 * [`endingFor`] is that field read. `malformed_call` is not ported: it patched
 * a missing protocol, and the protocol is here now.
 *
 * EVERY NAME BELOW IS REACHED BY A FOLD IN THIS PACKAGE. `answered`,
 * `truncated`, `refused` and `no calls` are the provider's finish signal read;
 * `round ceiling` is the tool-round guard in `step.js`; `failed` is a driver's
 * effect failure past the retry ceiling; `malformed reply` is a reply the
 * contract could not read; `empty completions` is the zero-output guard in
 * `retry.js`; `interrupted` is the boot that reads a checkpoint.
 *
 * WHAT IS STILL OWED. `pass ceiling`, `goal unmet`, `critic faulted` and
 * `unchecked` each rest on a fold this loop does not compute — passes, the
 * standing goal, the critic's verdict, the verify gate. They arrive with the
 * fold that earns them; naming an ending nothing can reach would be a
 * vocabulary describing a machine that is not here.
 * @module
 */

import { emit } from './effect.js'
import { FINISH_REASONS, idle } from './turn.js'

/** @typedef {import('./effect.js').Effect} Effect */
/** @typedef {import('./state.js').AgentState} AgentState */
/** @typedef {import('./turn.js').FinishReason} FinishReason */

/** The one ending fact. Payload: `{why, rounds, turnId}`. */
export const ENDED = 'core.ended'

/** The model answered. The turn's cheap exit, and the only ending after which there is a reply to read. */
export const ANSWERED = 'answered'

/** The turn used every round of tool calls its agent file allows. */
export const ROUND_CEILING = 'round ceiling'

/** The reply hit the output ceiling mid-sentence. Its own name because the act is to raise the ceiling or narrow the question, not to read the answer. */
export const TRUNCATED = 'truncated'

/** The model declined to answer. Not a failure of this build, and a person is owed the distinction. */
export const REFUSED = 'refused'

/** The provider failed the completion. The one ending that says nothing about the work. */
export const FAILED = 'failed'

/**
 * The provider said it was calling tools and named none. A contradiction rather
 * than an answer: acting on it as prose is how the predecessor turned every
 * malformed call into `main finished`.
 */
export const NO_CALLS = 'no calls'

/**
 * THE REPLY CARRIED NO ENDING SIGNAL — no calls to run and no `finish` to read
 * — and a reply that says neither is BROKEN rather than pending.
 *
 * It used to be recorded as a dropped fact, which left `awaiting: 'model'` set
 * and the turn waiting for a second reply that answers a call already made. A
 * deadline would have ended it eventually; waiting on a deadline for something
 * already known to be broken spends the person's time to learn nothing, so
 * malformed is an ENDING. The record says which turn and how many rounds it had
 * behind it, like every other ending.
 */
export const MALFORMED = 'malformed reply'

/**
 * TWO CONSECUTIVE ZERO-OUTPUT COMPLETIONS from the same model and the same
 * finish signal. The model is answering deterministically and the answer is
 * nothing, so asking a third time spends the person's money to receive the same
 * silence. Its own name, because the repair is to change the request or the
 * model and never to wait.
 */
export const STALLED = 'empty completions'

/**
 * A RELOAD LANDED IN THE MIDDLE OF THIS TURN and the work outstanding could
 * have changed something, so it was not re-issued. The one ending nothing in
 * the loop decided: it is what a boot says instead of leaving a turn in limbo
 * (`checkpoint.js`).
 */
export const INTERRUPTED = 'interrupted'

/**
 * THE ENDING SPELLED AS A TOOL, for a model with no native call API. Everything
 * such a model "says" is a call, so the answer is one too, and the turn ends
 * because the model SAID to end it rather than because no call could be read
 * out of its prose (Agent Zero's `response` tool, whose handler sets
 * `break_loop`). A model with native calls never needs it and never sees it.
 */
export const RESPOND = 'respond'

/** @type {Record<FinishReason, string>} */
const BY_FINISH = {
  stop: ANSWERED,
  tool_calls: NO_CALLS,
  length: TRUNCATED,
  content_filter: REFUSED,
  refusal: REFUSED,
  error: FAILED,
  // A PROVIDER THAT DID NOT SAY. Not a synonym for `stop`: the reply may be
  // complete, truncated or refused and nothing on the wire distinguishes them,
  // so the turn ends saying it does not know rather than claiming an answer.
  unknown: 'the provider did not say why it stopped',
}

/**
 * Which ending a call-less reply earned, from the signal the provider sent.
 *
 * THE SIGNAL IS A STRING FROM ANOTHER PACKAGE, and `FinishReason` is erased
 * before it gets here: OpenAI sends `content_filter`, Anthropic's whole set
 * (`end_turn`, `max_tokens`, `stop_sequence`, `tool_use`) matches no name
 * below. Indexing blind returned `undefined`, `JSON.stringify` then dropped the
 * `why` key from the ending record, and the turn ended with the log unable to
 * say why — the exact hole this file exists to close (I16). So the signal is
 * checked against the closed set and an unknown one is QUOTED BACK: a person
 * reading the ending sees the word the provider actually sent, which is the
 * only thing that tells them what to add.
 * @param {string} finish @returns {string}
 */
export function endingFor(finish) {
  const known = FINISH_REASONS.find((name) => name === finish)
  if (!known) return `unknown finish signal "${finish}"`
  return BY_FINISH[known]
}

/**
 * END THE TURN, AND SAY WHY. Every arm that ends one comes through here, so
 * what an ending CLEARS is written once and the reason is never optional.
 *
 * `stopping` clears with the rest: a stop ends one turn, not the next.
 * @param {AgentState} state @param {string} why @returns {{state: AgentState, effects: Effect[]}}
 */
export function endTurn(state, why) {
  const effect = emit({
    type: 'custom',
    kind: ENDED,
    payload: { why, rounds: state.toolRounds, turnId: state.turnId },
  })
  return { state: idle(state), effects: [effect] }
}

/**
 * Why the turn ended, out of the payload. An unreadable record says nothing
 * rather than guessing a reason — which is what every log written before this
 * fact existed reads as.
 * @param {unknown} payload @returns {string}
 */
export function endedWhy(payload) {
  const why = read(payload, 'why')
  return typeof why === 'string' ? why : ''
}

/** How many rounds of tool calls it had completed when it ended. @param {unknown} payload @returns {number} */
export function endedRounds(payload) {
  const rounds = read(payload, 'rounds')
  return typeof rounds === 'number' ? rounds : 0
}

/** @param {unknown} payload @param {string} key @returns {unknown} */
function read(payload, key) {
  if (typeof payload !== 'object' || payload === null) return undefined
  return /** @type {Record<string, unknown>} */ (payload)[key]
}
